driver/vllm/src/pie_driver_vllm/loader.py

```
@@ -219,24 +219,35 @@ def _debug_stage(msg: str) -> None:
         # Single-rank fallback. FileStore avoids picking a port (no contention).
+        #
+        # Use the gloo (CPU) backend, NOT nccl: on one GPU no GPU-side
+        # collectives ever run (vLLM's tensor_model_parallel_all_reduce
+        # short-circuits at world_size==1), but initializing an NCCL
+        # communicator + vLLM's device communicators here eagerly reserves
+        # ~2.8 GB of GPU memory (NCCL comm buffers + kernel modules) that is
+        # never used single-GPU. On small cards (e.g. a 4 GB RTX 3050 under
+        # WSL2) that overhead alone can leave no room for the KV cache and
+        # trips "Insufficient KV cache budget" (see issue #450). gloo keeps the
+        # group valid for CPU metadata broadcasts while touching no GPU memory.
         store_path = tempfile.mktemp(prefix="pie_vllm_singlerank_")
```

Ruff (0.15.20)

[error] Use of insecure and deprecated function (`mktemp`)

(S306)